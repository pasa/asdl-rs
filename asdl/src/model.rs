use std::fmt::Write;
use rustc_hash::FxHashMap;
use serde::Serialize;
use crate::parser;

#[derive(Serialize, Debug)]
pub struct Asdl {
    pub types: FxHashMap<String, Type>,
    pub prod_types: Vec<String>,
    pub sum_types: Vec<String>,
}

impl Asdl {
    pub(crate) fn new(root: &parser::Root) -> Self {
        let mut prod_types = Vec::new();
        let mut sum_types = Vec::new();
        for ty in root.types() {
            match ty {
                parser::Type::SumType(sty) => sum_types.push(sty.type_id().id().to_string()),
                parser::Type::ProdType(pty) => prod_types.push(pty.type_id().id().to_string()),
            }
        }
        let types = root.types().map(ty).map(|t| (t.id(), t)).collect();
        Asdl { types, prod_types, sum_types }
    }
}

#[derive(Serialize, Debug)]
#[serde(untagged)]
pub enum Type {
    SumType(SumType),
    ProdType(ProdType),
}

impl Type {
    fn id(&self) -> String {
        match self {
            Type::SumType(sty) => sty.id.clone(),
            Type::ProdType(pty) => pty.id.clone(),
        }
    }
}

impl From<SumType> for Type {
    fn from(n: SumType) -> Type {
        Type::SumType(n)
    }
}
impl From<ProdType> for Type {
    fn from(n: ProdType) -> Type {
        Type::ProdType(n)
    }
}

fn ty(ty: &parser::Type) -> Type {
    match ty {
        parser::Type::SumType(sty) => sum_type(sty).into(),
        parser::Type::ProdType(pty) => prod_type(pty).into(),
    }
}

#[derive(Serialize, Debug)]
pub struct SumType {
    id: String,
    constructors: Vec<Constructor>,
    attributes: Vec<Field>,
    is_prod_type: bool, //always false
}

impl SumType {
    fn new(id: String, constructors: Vec<Constructor>, attributes: Vec<Field>) -> Self {
        SumType { id, constructors, attributes, is_prod_type: false }
    }
}

fn sum_type(ty: &parser::SumType) -> SumType {
    let id = ty.type_id().id().to_string();
    let constructors = ty.constructors().map(constr).collect();
    let attributes = if let Some(attrs) = ty.attrs() {
        let mut names = FieldNames::default();
        attrs.fields().map(|f| field(f, &mut names)).collect()
    } else {
        vec![]
    };
    SumType::new(id, constructors, attributes)
}

#[derive(Serialize, Debug)]
pub struct Constructor {
    id: String,
    fields: Vec<Field>,
}

impl Constructor {
    fn new(id: String, fields: Vec<Field>) -> Self {
        Constructor { id, fields }
    }
}

fn constr(c: &parser::Constr) -> Constructor {
    let mut names = FieldNames::default();
    let fields: Vec<Field> = c.fields().map(|f| field(f, &mut names)).collect();
    Constructor::new(c.id().id().to_string(), fields)
}

#[derive(Serialize, Debug)]
pub struct ProdType {
    id: String,
    fields: Vec<Field>,
    is_prod_type: bool, //always true
}

impl ProdType {
    fn new(id: String, fields: Vec<Field>) -> Self {
        ProdType { id, fields, is_prod_type: true }
    }
}

fn prod_type(ty: &parser::ProdType) -> ProdType {
    let mut names = FieldNames::default();
    let fields: Vec<Field> = ty.fields().map(|f| field(f, &mut names)).collect();
    ProdType::new(ty.type_id().id().to_string(), fields)
}

#[derive(Serialize, Debug)]
pub struct Field {
    id: String,
    type_id: String,
    is_single: bool,
    is_option: bool,
    is_sequence: bool,
}

impl Field {
    fn single(id: String, type_id: String) -> Self {
        Field { id, type_id, is_single: true, is_option: false, is_sequence: false }
    }

    fn option(id: String, type_id: String) -> Self {
        Field { id, type_id, is_single: false, is_option: true, is_sequence: false }
    }

    fn sequence(id: String, type_id: String) -> Self {
        Field { id, type_id, is_single: false, is_option: false, is_sequence: true }
    }
}

fn field(f: &parser::Field, names: &mut FieldNames) -> Field {
    match f {
        parser::Field::Single(f) => {
            Field::single(names.get_or_generate(f.id(), f.type_id()), f.type_id().id().to_string())
        }
        parser::Field::Opt(f) => {
            Field::option(names.get_or_generate(f.id(), f.type_id()), f.type_id().id().to_string())
        }
        parser::Field::Sequence(f) => Field::sequence(
            names.get_or_generate(f.id(), f.type_id()),
            f.type_id().id().to_string(),
        ),
    }
}

#[derive(Default)]
struct FieldNames {
    names_indexes: FxHashMap<String, u32>,
}

impl FieldNames {
    fn get_or_generate(&mut self, id: &Option<parser::Id>, type_id: &parser::TypeId) -> String {
        match id {
            Option::Some(id) => id.id().to_string(),
            Option::None => {
                let type_id = type_id.id();
                let index = self.names_indexes.entry(type_id.to_string()).or_insert(0);
                let res = if *index == 0 {
                    type_id.to_string()
                } else {
                    let mut buf = String::new();
                    write!(&mut buf, "{}{}", type_id, index).unwrap();
                    buf
                };
                *index += 1;
                res
            }
        }
    }
}
