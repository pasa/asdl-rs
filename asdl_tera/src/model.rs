use std::collections::HashMap;
use serde::Serialize;
use asdl;
use super::Result;

#[derive(Serialize, Debug)]
pub struct Asdl {
    pub types: HashMap<String, Type>,
    pub prod_types: Vec<String>,
    pub sum_types: Vec<String>,
    pub comments: Vec<String>,
}

impl Asdl {
    pub fn parse(asdl: &str) -> Result<Asdl> {
        Ok(Asdl::new(asdl::Asdl::parse(asdl)?))
    }
}

#[derive(Serialize, Debug)]
#[serde(untagged)]
pub enum Type {
    SumType(SumType),
    ProdType(ProdType),
}

impl From<SumType> for Type {
    fn from(n: SumType) -> Type {
        Type::SumType(n)
    }
}
impl From<ProdType> for Type {
    fn from(n: ProdType) -> Type {
        Type::ProdType(n)
    }
}

#[derive(Serialize, Debug)]
pub struct SumType {
    pub id: String,
    pub constructors: Vec<Constructor>,
    pub attributes: Vec<Field>,
    pub is_prod_type: bool, //always false
    pub comments: Vec<String>,
}

#[derive(Serialize, Debug)]
pub struct Constructor {
    pub id: String,
    pub fields: Vec<Field>,
    pub comments: Vec<String>,
}

#[derive(Serialize, Debug)]
pub struct ProdType {
    pub id: String,
    pub fields: Vec<Field>,
    pub is_prod_type: bool, //always true
    pub comments: Vec<String>,
}

#[derive(Serialize, Debug)]
pub struct Field {
    pub id: String,
    pub type_id: String,
    pub is_required: bool,
    pub is_optional: bool,
    pub is_repeated: bool,
}
