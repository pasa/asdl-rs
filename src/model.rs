use std::fmt::Write;

use rustc_hash::FxHashMap;
use heck::MixedCase;
use linked_hash_set::LinkedHashSet;
use serde::Serialize;

use crate::parser;

#[derive(Serialize, Debug)]
pub struct Asdl {
    sum_types: Vec<SumType>,
    prod_types: Vec<ProdType>,
}

#[derive(Serialize, Debug)]
pub(crate) struct SumType {
    id: String,
    constructors: Vec<Constructor>,
    attributes: Vec<Field>,
}

#[derive(Serialize, Debug)]
pub(crate) struct Constructor {
    id: String,
    fields: Vec<Field>,
}

#[derive(Serialize, Debug, Clone, Hash, Eq, PartialEq)]
pub(crate) struct ProdType {
    id: String,
    fields: Vec<Field>,
}

#[derive(Serialize, Debug, Clone, Hash, Eq, PartialEq)]
pub(crate) struct Field {
    id: String,
    type_id: String,
    is_single: bool,
    is_option: bool,
    is_sequence: bool,
}

impl Field {
    fn single(id: String, type_id: String) -> Field {
        Field { id, type_id, is_single: true, is_option: false, is_sequence: false }
    }

    fn option(id: String, type_id: String) -> Field {
        Field { id, type_id, is_single: false, is_option: true, is_sequence: false }
    }

    fn sequence(id: String, type_id: String) -> Field {
        Field { id, type_id, is_single: false, is_option: false, is_sequence: true }
    }
}

impl Asdl {
    pub(crate) fn new(root: &parser::Root) -> Self {
        let mut res = Asdl { prod_types: vec![], sum_types: vec![] };
        for d in root.types() {
            match d.kind() {
                parser::TypeKind::SumType(t) => {
                    res.sum_types.push(sum_type(t));
                }
                parser::TypeKind::ProdType(t) => {
                    res.prod_types.push(prod_type(t));
                }
            }
        }
        res
    }
}

fn sum_type(node: &parser::SumType) -> SumType {
    let id = node.type_id().text().to_string();
    let constructors = node.constructors().map(constructor).collect();
    let attributes = if let Some(attrs) = node.attrs() {
        let mut names = FieldNames::default();
        attrs.fields().map(|f| field(f, &mut names)).collect()
    } else {
        vec![]
    };
    SumType { id, constructors, attributes }
}

fn constructor(node: &parser::Constr) -> Constructor {
    let id = node.id().text().to_string();
    let mut names = FieldNames::default();
    let fields: Vec<Field> = node.fields().map(|f| field(f, &mut names)).collect();
    Constructor { id, fields }
}

fn field(node: &parser::Field, names: &mut FieldNames) -> Field {
    match node.kind() {
        parser::FieldKind::Single(f) => Field::single(
            names.get_or_generate(f.id(), f.type_id()),
            f.type_id().text().to_string(),
        ),
        parser::FieldKind::Opt(f) => Field::option(
            names.get_or_generate(f.id(), f.type_id()),
            f.type_id().text().to_string(),
        ),
        parser::FieldKind::Sequence(f) => Field::sequence(
            names.get_or_generate(f.id(), f.type_id()),
            f.type_id().text().to_string(),
        ),
    }
}

fn prod_type(node: &parser::ProdType) -> ProdType {
    let id = node.type_id().text().to_string();
    let mut names = FieldNames::default();
    let fields: Vec<Field> = node.fields().map(|f| field(f, &mut names)).collect();
    ProdType { id, fields }
}

#[derive(Default)]
struct FieldNames {
    names_indexes: FxHashMap<String, u32>,
}

impl FieldNames {
    fn get_or_generate(&mut self, id: Option<&parser::Id>, type_id: &parser::TypeId) -> String {
        match id {
            Option::Some(id) => id.text().to_string(),
            Option::None => {
                let type_id = type_id.text();
                let index = self.names_indexes.entry(type_id.to_string()).or_insert(0);
                let res = if *index == 0 {
                    type_id.to_string()
                } else {
                    let mut buf = String::new();
                    write!(&mut buf, "{}{}", type_id, index).unwrap();
                    buf
                };
                *index += 1;
                res
            }
        }
    }
}
