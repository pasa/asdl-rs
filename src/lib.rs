mod model;
mod parser;

use tera::*;
use crate::model::*;
use failure::format_err;
use heck::{CamelCase, ShoutySnakeCase, SnakeCase, MixedCase};

pub fn generate(asdl: &str, template: &str) -> String {
    let root: &parser::Root = &parser::parse(&asdl);
    let model = Asdl::from(root);
    let mut tera = Tera::default();
    tera.register_filter("camel", |arg, _| Ok(arg.as_str().unwrap().to_camel_case().into()));
    tera.register_filter("snake", |arg, _| Ok(arg.as_str().unwrap().to_snake_case().into()));
    tera.register_filter("mixed", |arg, _| Ok(arg.as_str().unwrap().to_mixed_case().into()));
    tera.register_filter("SCREAM", |arg, _| {
        Ok(arg.as_str().unwrap().to_shouty_snake_case().into())
    });
    tera.add_raw_template("_src", &template)
        .map_err(|e| format_err!("template parsing error: {:?}", e))
        .unwrap();
    tera.render("_src", &model)
        .map_err(|e| format_err!("template rendering error: {:?}", e))
        .unwrap()
}

#[cfg(test)]
mod tests {
    use crate::*;
    use insta::assert_snapshot_matches;
    use insta::assert_debug_snapshot_matches;

    #[test]
    fn simple_successful_test() {
        let asdl = r"
            stm = Compound(stm s1, stm* s2)
                | Single(stm)
            noFileds = One | Two | Tree
            prodType = (noFileds? f, stm s1)
            ";
        let root: &parser::Root = &parser::parse(&asdl);
        assert_snapshot_matches!("simple_successful_test_syntax", root.debug_dump());
        let model = Asdl::from(root);
        assert_debug_snapshot_matches!("simple_successful_test_model", model)
    }
}
