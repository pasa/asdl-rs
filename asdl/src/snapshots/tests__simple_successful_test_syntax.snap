---
created: "2019-08-15T11:58:41.798987386Z"
creator: insta@0.8.1
source: src/lib.rs
expression: root
---
Root {
    types: [
        SumType(
            SumType {
                type_id: TypeId(
                    "stm",
                ),
                constructors: [
                    Constr {
                        id: ConstrId(
                            "Compound",
                        ),
                        fields: [
                            Single(
                                Single {
                                    type_id: TypeId(
                                        "stm",
                                    ),
                                    id: Some(
                                        Id(
                                            "s1",
                                        ),
                                    ),
                                },
                            ),
                            Sequence(
                                Sequence {
                                    type_id: TypeId(
                                        "stm",
                                    ),
                                    id: Some(
                                        Id(
                                            "s2",
                                        ),
                                    ),
                                },
                            ),
                        ],
                    },
                    Constr {
                        id: ConstrId(
                            "Single",
                        ),
                        fields: [
                            Single(
                                Single {
                                    type_id: TypeId(
                                        "stm",
                                    ),
                                    id: None,
                                },
                            ),
                        ],
                    },
                ],
                attrs: None,
            },
        ),
        SumType(
            SumType {
                type_id: TypeId(
                    "noFileds",
                ),
                constructors: [
                    Constr {
                        id: ConstrId(
                            "One",
                        ),
                        fields: [],
                    },
                    Constr {
                        id: ConstrId(
                            "Two",
                        ),
                        fields: [],
                    },
                    Constr {
                        id: ConstrId(
                            "Three",
                        ),
                        fields: [],
                    },
                ],
                attrs: None,
            },
        ),
        ProdType(
            ProdType {
                type_id: TypeId(
                    "prodType",
                ),
                fields: [
                    Opt(
                        Opt {
                            type_id: TypeId(
                                "noFileds",
                            ),
                            id: Some(
                                Id(
                                    "f",
                                ),
                            ),
                        },
                    ),
                    Single(
                        Single {
                            type_id: TypeId(
                                "stm",
                            ),
                            id: Some(
                                Id(
                                    "s1",
                                ),
                            ),
                        },
                    ),
                ],
            },
        ),
    ],
}
